godot-core/src/builtin/array.rs

```diff
@@ -363,6 +373,20 @@ impl<T: GodotType> Array<T> {
             }
         }
     }
+
+    /// Creates a new reference to the data in this array.
+    ///
+    /// This does not ensure the Array's type matches what we statically expect.
+    fn clone_unchecked(&self) -> Self {
+        // SAFETY: `self` is a valid array, since we have a reference that keeps it alive.
+        unsafe {
+            Self::from_sys_init(|self_ptr| {
+                let ctor = ::godot_ffi::builtin_fn!(array_construct_copy);
+                let args = [self.sys_const()];
+                ctor(self_ptr, args.as_ptr());
+            })
+        }
+    }
```

Comment:
I dont think this is actually a safety invariant, it's just a correctness invariant. As i don't believe anything in `Array` actually relies on the values stored in the array to be `T`, and anything that expects it to be will panic/error when it's not. If this actually is unsafe then i'd need to rewrite some stuff. Maybe @ttencate knows for sure?

We can refer to the `assume_type` method here, whose safety doc says:

```rs
In and of itself, calling this does not result in undefined behavior. However:
- If `T` is not `Variant`, the returned array should not be written to, because the runtime type check may fail.
- If `U` is not `Variant`, the returned array should not be read from, because conversion from variants may fail.

In the current implementation, both cases will produce a panic rather than undefined
behavior, but this should not be relied upon.
```

I guess not marking this as `unsafe` would be a commitment to the current behavior, meaning `assume_type` should be changed to be a safe function. Which wouldn't be the worst, we could in the future provide `*_unchecked` methods as alternatives in cases where we currently panic?